docs/design/in-doc-search-ui.rst

```diff
@@ -0,0 +1,183 @@
+In Doc Search UI
+================
+
+Giving readers the ability to easily search the information
+that they are looking for is important for us.
+We have already upgraded to the latest version of `Elasticsearch`_ and
+we plan to implement `search as you type` feature for all the documentations hosted by us.
+It will be designed to provide instant results as soon as the user starts
+typing in the search bar with a clean and minimal frontend.
+This design document aims to provides the details of it.
+This is a GSoC'19 project.
+
+.. warning::
+
+    This design document details future features that are **not yet implemented**.
+    To discuss this document, please get in touch in the `issue tracker`_.
+
+
+The final result may look something like this:
+
+.. figure:: ../_static/images/design-docs/in-doc-search-ui/in-doc-search-ui-demo.gif
+    :align: center
+    :target: ../_static/images/design-docs/in-doc-search-ui/in-doc-search-ui-demo.gif
+
+    Short demo
+
+
+Goals And Non-Goals
+-------------------
+
+Project Goals
+++++++++++++++
+
+* Support a search-as-you-type/autocomplete interface.
+* Support across all (or virtually all) Sphinx themes.
+* Support for the JavaScript user experience down to IE11 or graceful degradation where we can't support it.
+* Project maintainers should have a way to opt-in/opt-out of this feature.
+* (Optional) Project maintainers should have the flexibility to change some of the styles using custom CSS and JS files.
+
+Non-Goals
+++++++++++
+
+* For the initial release, we are targeting only Sphinx documentations
+  as we don't index MkDocs documentations to our Elasticsearch index.
```

Comment:
Of note, supporting _all_ Sphinx themes will also make it much easier for us to support mkdocs. A large part of the issue with around the UI/UX and theme integration, so we should hopefully be able to work around that now :)